[jvm-packages] update local dev build process

thesuperzapper opened the pull request:
This PR changes the dev (non-ci) build process for xgboost4j to use a locally defined Dockerfile. The main purpose of this it to upgrade to gcc 5.X in line with PR #4538, as the old container can no longer build XGBoost.

thesuperzapper commented:
@CodingCat, this is that thing we discussed about updating the Dockerfile for local builds, thoughts?

CodingCat commented:
I didn't understand why we need such changes to enable new gcc, isn't it just require an update of docker image?

thesuperzapper commented:
Its better if we have the docker image in the codebase so people can see what image they are using to build. 

The image I have used here is pretty much the [same one we use in CI for JVM](https://github.com/dmlc/xgboost/blob/master/tests/ci_build/Dockerfile.jvm).

CodingCat commented:
> Its better if we have the docker image in the codebase so people can see what image they are using to build.
> 
> The image I have used here is pretty much the [same one we use in CI for JVM](https://github.com/dmlc/xgboost/blob/master/tests/ci_build/Dockerfile.jvm).

1. I do not think we should rebuild the image by installing everything for every time, there are many many times I am using docker to run unit test in linux (for the different behavior of linux/mac)...it will waste a lot of time when we are in the dev cycle

2. even we install everything, we should not duplicate docker files but rather using a bash script to refer to the one used by CI

thesuperzapper commented:
@CodingCat are you referring to the MVN download, or the building/tagging of the Dockerfile?

* The new build.sh script asks if you want to rebuild the docker file.
* In terms of pointing at the CI one, there are a few changes, like that we don't use ENTRYPOINT, or gosu. (And isn't it better if this is slightly separate so changes to the CI image don't break what we are doing?)
* I agree that we should look into stopping the MVN download each time, we can probably do this with a volume mount.

CodingCat commented:
> @CodingCat are you referring to the MVN download, or the building/tagging of the Dockerfile?
> 
> * The new build.sh script asks if you want to rebuild the docker file.
> * In terms of pointing at the CI one, there are a few changes, like that we don't use ENTRYPOINT, or gosu. (And isn't it better if this is slightly separate so changes to the CI image don't break what we are doing?)
> * I agree that we should look into stopping the MVN download each time, we can probably do this with a volume mount.

I am talking about yum install part

thesuperzapper commented:
@CodingCat with this PR, if you run `./dev/build.sh` it will ask you, "do you want to rebuild/tag the Docker image", if you say "no" it will just use a locally cached version (if available).

However, I agree that we should stop MVN trying to download the jars every time. (I will look into mounting the users home folder)

CodingCat commented:
what's the glibc version used in this PR?

thesuperzapper commented:
As it's based off centos6 it uses `ldd (GNU libc) 2.12`

CodingCat commented on jvm-packages/dev/change_version.sh:
```diff
@@ -1,27 +1,19 @@
 #
-# (Yizhi) This is mainly inspired by the script in apache/spark.
-# I did some modificaiton to get it with our project.
-# (Nan) Modified from MxNet
```
please keep these 3 lines

CodingCat commented on jvm-packages/dev/change_version.sh:
```diff
@@ -1,27 +1,19 @@
-# The ASF licenses this file to You under the Apache License, Version 2.0
-# (the "License"); you may not use this file except in compliance with
-# the License.  You may obtain a copy of the License at
+# Copyright (c) 2019 by Contributors
```
don't add years to avoid keep changing it

thesuperzapper replied:
I am fairly sure you need to keep the year of release for legal reasons.

CodingCat replied:
Just look at the code base how many 2014 2017 2018 have been there

thesuperzapper replied:
When I look through the project I see lots of examples where they have put the year with copyright deceleration.
* https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j/src/main/java/ml/dmlc/xgboost4j/java/DMatrix.java#L2
* https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostRegressor.scala#L2

Its not safe to remove the copyright deceleration.

CodingCat replied:
those should be removed, I was not aware that our scala style checker is actually a bit off

https://github.com/dmlc/xgboost/blob/master/jvm-packages/scalastyle-config.xml#L52 this should be removed and we should follow standard apache license header which does not include year info
https://www.apache.org/legal/src-headers.html#headers

CodingCat commented on jvm-packages/dev/package.sh:
```diff
@@ -0,0 +1,26 @@
+#!/usr/bin/env bash
```
rename this file to package-linux.sh

CodingCat commented on jvm-packages/dev/build.sh:
```diff
@@ -1,21 +1,37 @@
+      [Nn]* ) break;;
+      * ) echo "Invalid Response: please answer [y]es or [n]o";;
+  esac
+done
```
let's replace all interactive questions with the passed in arguments

thesuperzapper replied:
I think it makes it less usable when quickly developing local changes. 

That is, you will have to type something like `--skip-build` every time you don't want to rebuild the Dockerfile. (This also means we need to write documentation about it.)

I will defer to your opinion however.

CodingCat replied:
Yes, by default, we should keep unit test and skip rebuild image

thesuperzapper replied:
Also, the image wont rebuild unless the Dockerfile has changed, so do we even need to have this flag?

The issue with having skip building on default, is that if you have never built it before, the `docker run` step will fail, because the image clearly doesn't exist on your local cache.

CodingCat replied:
Then why you have this interactive question?

thesuperzapper replied:
It originally forced a rebuild by clearing the cache. 

Do you want me to introduce that behaviour with flag to force the rebuild, and default behaviour to NOOP if no changes are detected in Dockerfile?

thesuperzapper replied:
I have removed this, it will now attempt to rebuild every time, and only do something if you change the Dockerfile, or have never built it locally before.

CodingCat commented on jvm-packages/dev/Dockerfile:
```diff
@@ -0,0 +1,47 @@
+FROM centos:6
+
+# Install all basic requirements
+RUN \
```
after taking a second look, I think it might be better to keep a separate copy of Dockerfile here to ensure that we have a consistent release env

CodingCat commented on jvm-packages/dev/build.sh:
```diff
@@ -1,21 +1,37 @@
 #!/usr/bin/env bash
+#
+# Copyright (c) 2019 by Contributors
```
remove this to build-linux.sh

thesuperzapper commented on jvm-packages/dev/package.sh:
```diff
@@ -0,0 +1,26 @@
+cd jvm-packages
+
+while true; do
+  read -p  "Do you want to skip unit tests?" yn
```
@CodingCat, should this one be replaced with allowing the user to pass `-DskipTests`?

E.g. `./dev/build-linux.sh -DskipTests`

We should only need this if you plan to run this without user present.

thesuperzapper commented:
@CodingCat I have made some changes:
* The maven cache is now mounted inside the ./dev folder, meaning it will not re-download the jars every time. (This folder has been added to .gitignore)
* I have renamed the files you asked.
* I have re-added the licence comments to change_version.sh
* I have removed the prompt to rebuild the docker image, (It will now just build if the Dockerfile changes, or has never been build locally before)
* I have stopped build-linux.sh creating a build.log

Question:
* Assuming you don't want the interactive prompt for skipping tests, what flag for build-linux.sh do you want me to create? (`--skip-tests`, `-DskipTests`, etc?)

thesuperzapper commented on jvm-packages/dev/change_version.sh:
```diff
@@ -19,9 +15,9 @@
 #
 # (Yizhi) This is mainly inspired by the script in apache/spark.
-# I did some modificaiton to get it with our project.
+# I did some modifications to get it with our project.
```
@CodingCat should I leave the spelling mistake for benefit of git history?

CodingCat replied:
fix typos is fine

thesuperzapper commented:
@CodingCat I have now made it so that tests can be skipped using `--skip-tests`, and maven caches the jars locally in `./dev/.m2`.

The way you use the build is: `./dev/build-linux.sh [--skip-tests]`

thesuperzapper commented:
@CodingCat I have attempted to make a `build-linux.cmd` for windows docker environment, but this can be a bit flaky, as `package-linux.sh` sometimes has its line endings change to Windows when you git clone. 

Should we COPY the `package-linux.sh` into the Dockerfile? Or are you happy having it referenced from a local volume mount?

thesuperzapper commented:
@CodingCat don't worry, I fixed it with a .gitattributes file.

trams commented:
I tried https://github.com/dmlc/xgboost/pull/4640/commits/7cf9f941f74e37982991fb8a5730d1e7e2ab1751 in my locally checked out xgboost folder and build-linux.sh failed with

`CMake Error: The current CMakeCache.txt directory /xgboost/build/CMakeCache.txt is different than the directory /home/projects/xgboost/build where CMakeCache.txt was created. This may result in binaries being created in the wrong place. If you are not sure, reedit the CMakeCache.txt
CMake Error: The source "/xgboost/CMakeLists.txt" does not match the source "/home/projects/xgboost/CMakeLists.txt" used to generate cache.  Re-run cmake with a different source directory.
building Java wrapper
cd ..
mkdir -p build
cd build
cmake .. -DUSE_OPENMP:BOOL=ON -DUSE_HDFS:BOOL=OFF -DUSE_AZURE:BOOL=OFF -DUSE_S3:BOOL=OFF -DUSE_CUDA:BOOL=OFF -DJVM_BINDINGS:BOOL=ON
Traceback (most recent call last):
  File "create_jni.py", line 92, in <module>
    run("cmake .. " + " ".join(args) + maybe_generator)
  File "create_jni.py", line 51, in run
    subprocess.check_call(command, shell=True, **kwargs)
  File "/opt/python/lib/python3.7/subprocess.py", line 341, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'cmake .. -DUSE_OPENMP:BOOL=ON -DUSE_HDFS:BOOL=OFF -DUSE_AZURE:BOOL=OFF -DUSE_S3:BOOL=OFF -DUSE_CUDA:BOOL=OFF -DJVM_BINDINGS:BOOL=ON' returned non-zero exit status 1.
`

I think it is because I have already build it locally

I'll retry it on the clean clone

thesuperzapper commented:
@trams this usually happens when you attempt to build the project with your local machine, (which is almost certainly not running centos6). 

Please try deleting the `./build` directory under the project root.

trams commented:
Thank you for reply.

Should we check for this folder or should we use another folder name for in-Docker build?

thesuperzapper commented:
@CodingCat do you want me to rename the cmake build directory for JVM builds? 

How about: `./build_jvm`?

CodingCat commented:
> @CodingCat do you want me to rename the cmake build directory for JVM builds?
> 
> How about: `./build_jvm`?

is it really necessary? I personally think it will make the process error-prone. Clean everything after building for another arch is not something "too strict"....

Though I didn't verify, if we have different output folder, are we even able to run linux & mac build in parallel? and if that's the case, are we messing up the object files? even not, I would choose to keep the build step as strict as possible to avoid any unexpected behavior

thesuperzapper commented:
@CodingCat I have squashed the commits, is this good to merge now?

CodingCat commented:
LGTM

trams commented:
Sorry for polluting this closed pull request.

I rerun jvm dev scripts on my linux dev machine again and I realised what kind of problem did I have.
The problem is that all files created inside the docker are owned by root.

So if I have a folder which I develop & build locally then I use this scripts to build a jar then I cannot go back to developing locally without manually cleaning all artifacts cause they are all owned by root

As far as I understand you should not have this problem on OS X or Windows

Why do I want to use these scripts if I already have a linux machine? The reason is I need to build a native shared library inside CentOS 6 docker otherwise I won't be able to use xgboost-spark on my cluster cause this cluster has an old OS with an old libc

That being said there are some work around (mainly this one https://www.jujens.eu/posts/en/2017/Jul/02/docker-userns-remap/) but I think after applying it I won't be able to use ci_build scripts

What do you think if I would modify the Docker container and the scripts to use entrypoint.sh like all other CI build scripts do?
